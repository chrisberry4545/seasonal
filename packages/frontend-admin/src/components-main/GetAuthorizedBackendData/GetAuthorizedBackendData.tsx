import React, { FC, useState, useEffect } from 'react';
import { IDataFormConfigProps } from '../DataForm/DataForm';
import { LoadingSpinner } from '@chrisb-dev/seasonal-shared-frontend-components';
import './GetAuthorizedBackendData.scss';

export interface IGetAuthorizedBackendDataProps<T> {
  items: T;
  reload?: () => void;
  updateMethod?: (item: T) => Promise<T>;
  buttonText?: string;
  additionalConfig?: IDataFormConfigProps<T>;
}

interface IGetAuthorizedBackendDataState<T> {
  isLoading: boolean;
  items: T | null;
  error: string | null;
}

export function GetAuthorizedBackendData<T>({
  InnerComponent,
  requestDataMethod,
  updateMethod,
  buttonText,
  additionalConfig
}: {
  InnerComponent: FC<IGetAuthorizedBackendDataProps<T>>,
  requestDataMethod: () => Promise<T>,
  updateMethod?: (item: T) => Promise<T>,
  buttonText?: string,
  additionalConfig?: IDataFormConfigProps<T>
}) {

  const [state, setState] = useState<IGetAuthorizedBackendDataState<T>>({
    error: null,
    isLoading: true,
    items: null
  });

  const requestNewData = () => {
    requestDataMethod()
      .then((items) => setState({
        error: null,
        isLoading: false,
        items
      })).catch((error) => setState({
        error: error.message,
        isLoading: false,
        items: null
      }));
  };

  useEffect(requestNewData, []);

  return (
    <div>
      {
        state.isLoading
          ? <div className='c-get-authorized-backend-data__loading'>
            <LoadingSpinner />
          </div>
          : <div>
            {
              !state.error && state.items
                ? <InnerComponent
                    items={state.items}
                    reload={requestNewData}
                    updateMethod={updateMethod}
                    buttonText={buttonText}
                    additionalConfig={additionalConfig}
                  />
                : <div>{state.error}</div>
            }
          </div>
      }
    </div>
  );
}
