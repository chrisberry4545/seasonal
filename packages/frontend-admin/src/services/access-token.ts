import { getStorage } from '../consts';

export const setAccessToken = (token: string) => {
  getStorage().accessToken = token;
};

export const getAccessTokenHeaders = (): HeadersInit => {
  const token = getStorage().accessToken;
  return token
  ? {
    Authorization: `Bearer ${token}`
  }
  : {};
};
