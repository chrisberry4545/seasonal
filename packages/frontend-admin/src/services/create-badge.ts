import { IBadge } from '@chrisb-dev/seasonal-shared-models';
import { makeAuthorizedRequest } from './make-authorized-request';
import { BADGE_URL } from '../config';

export const createBadge = async (badge: IBadge): Promise<IBadge> =>
  makeAuthorizedRequest<IBadge>(BADGE_URL, {
    body: JSON.stringify(badge),
    method: 'POST'
  });
