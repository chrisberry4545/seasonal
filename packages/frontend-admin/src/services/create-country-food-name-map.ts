import { ICountryFoodNameMap } from '@chrisb-dev/seasonal-shared';
import { makeAuthorizedRequest } from './make-authorized-request';
import { COUNTRY_FOOD_NAME_MAP_URL } from '../config';

export const createCountryFoodNameMap = async (
  country: ICountryFoodNameMap
): Promise<ICountryFoodNameMap> =>
  makeAuthorizedRequest<ICountryFoodNameMap>(COUNTRY_FOOD_NAME_MAP_URL, {
    body: JSON.stringify(country),
    method: 'POST'
  });
