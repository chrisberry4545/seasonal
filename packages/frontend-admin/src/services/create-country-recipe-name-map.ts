import { ICountryRecipeNameMap } from '@chrisb-dev/seasonal-shared-models';
import { makeAuthorizedRequest } from './make-authorized-request';
import { COUNTRY_RECIPE_NAME_MAP_URL } from '../config';

export const createCountryRecipeNameMap = async (
  country: ICountryRecipeNameMap
): Promise<ICountryRecipeNameMap> =>
  makeAuthorizedRequest<ICountryRecipeNameMap>(COUNTRY_RECIPE_NAME_MAP_URL, {
    body: JSON.stringify(country),
    method: 'POST'
  });
