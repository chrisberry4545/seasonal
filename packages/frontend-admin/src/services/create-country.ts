import { ICountry } from '@chrisb-dev/seasonal-shared-models';
import { makeAuthorizedRequest } from './make-authorized-request';
import { COUNTRY_URL } from '../config';

export const createCountry = async (country: ICountry): Promise<ICountry> =>
  makeAuthorizedRequest<ICountry>(COUNTRY_URL, {
    body: JSON.stringify(country),
    method: 'POST'
  });
