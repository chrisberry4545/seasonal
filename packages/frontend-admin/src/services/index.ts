export * from './set-access-token';

export * from './create-badge';

export * from './create-country-food-name-map';

export * from './create-country-recipe-name-map';

export * from './create-country';

export * from './create-food';

export * from './create-recipe';

export * from './create-region-food-season-map';

export * from './create-region';

export * from './create-user';

export * from './delete-badge';

export * from './delete-country-food-name-map';

export * from './delete-country-recipe-name-map';

export * from './delete-country';

export * from './delete-food';

export * from './delete-recipe';

export * from './delete-region-food-season-map';

export * from './delete-region';

export * from './delete-user';

export * from './get-all-badges';

export * from './get-all-countries';

export * from './get-all-country-food-name-map';

export * from './get-all-country-recipe-name-map';

export * from './get-all-food';

export * from './get-all-recipes';

export * from './get-all-region-food-season-map';

export * from './get-all-regions';

export * from './get-all-users';

export * from './get-one-badge';

export * from './get-one-country';

export * from './get-one-country-food-name-map';

export * from './get-one-region-food-season-map';

export * from './get-one-country-recipe-name-map';

export * from './get-one-food';

export * from './get-one-recipe';

export * from './get-one-region';

export * from './get-one-user';

export * from './login-request';

export * from './make-authorized-request';

export * from './update-badge';

export * from './update-country';

export * from './update-country-food-name-map';

export * from './update-region-food-season-map';

export * from './update-country-recipe-name-map';

export * from './update-food';

export * from './update-recipe';

export * from './update-region';

export * from './update-user';
