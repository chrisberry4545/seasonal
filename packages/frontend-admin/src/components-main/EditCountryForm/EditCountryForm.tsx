import React, { FC } from 'react';
import {
  GetAuthorizedBackendData
} from '../GetAuthorizedBackendData/GetAuthorizedBackendData';
import { ICountry } from '@chrisb-dev/seasonal-shared';
import { useParams } from 'react-router-dom';
import { getSingleCountry, updateCountry } from '../../services';
import { BaseFormCountry } from '../BaseFormCountry/BaseFormCountry';
import { FormLayout } from '../../components-layouts';

export const EditCountryForm: FC<{}> = () => {
  const { id } = useParams();
  const CreatedComponent = GetAuthorizedBackendData<ICountry>(
    BaseFormCountry,
    () => {
      return getSingleCountry(id as string);
    },
    updateCountry
  );
  return (
    <FormLayout title='Edit Country'>
      <CreatedComponent />
    </FormLayout>
  );
};
