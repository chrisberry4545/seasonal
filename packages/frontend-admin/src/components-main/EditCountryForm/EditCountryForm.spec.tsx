
import React from 'react';
import { EditCountryForm } from './EditCountryForm';
import { shallow, ShallowWrapper } from 'enzyme';
import { ICountry } from '@chrisb-dev/seasonal-shared-models';
import * as services from '../../services';
import { GetAuthorizedBackendData } from '../GetAuthorizedBackendData/GetAuthorizedBackendData';

jest.mock('../../services', () => ({
  getOneCountry: jest.fn(),
  updateCountry: jest.fn()
}));
jest.mock('../../components-layouts', () => ({
  LayoutWithTitle: () => 'LayoutWithTitle'
}));
jest.mock('../GetAuthorizedBackendData/GetAuthorizedBackendData', () => ({
  GetAuthorizedBackendData: () => 'GetAuthorizedBackendData'
}));
jest.mock('../BaseFormCountry/BaseFormCountry', () => ({
  BaseFormCountry: () => 'BaseFormCountry'
}));
jest.mock('react-router-dom', () => ({
  useParams: jest.fn().mockReturnValue({ id: 'id' })
}));

describe('<EditCountryForm />', () => {
  let wrapper: ShallowWrapper;
  let mockGetOne: jest.SpyInstance;
  let mockUpdate: jest.SpyInstance;
  const item = {
    name: 'name'
  } as ICountry;

  beforeEach(() => {
    mockGetOne = jest.spyOn(services, 'getOneCountry')
      .mockResolvedValue(item);
    mockUpdate = jest.spyOn(services, 'updateCountry')
      .mockResolvedValue({} as ICountry);
    wrapper = shallow(
      <EditCountryForm />
    );
  });

  test('renders correctly', () => expect(wrapper).toMatchSnapshot());

  test('can get the item data', async () => {
    await wrapper.find(GetAuthorizedBackendData).props().requestDataMethod();
    expect(mockGetOne).toHaveBeenCalledWith('id');
  });

  test('can update the item', () => {
    const update = wrapper.find(GetAuthorizedBackendData).props().updateMethod;
    if (update) {
      update(item);
    }
    expect(mockUpdate).toHaveBeenCalledWith(item);
  });

});
