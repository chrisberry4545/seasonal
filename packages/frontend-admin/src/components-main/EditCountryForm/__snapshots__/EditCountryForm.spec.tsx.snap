// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditCountryForm /> renders correctly 1`] = `
<LayoutWithTitle
  title="Edit Country"
>
  <GetAuthorizedBackendData
    InnerComponent={[Function]}
    requestDataMethod={[Function]}
    updateMethod={[MockFunction]}
  />
</LayoutWithTitle>
`;
